kratos/input_output/vtk_output.cpp

```diff
@@ -124,12 +124,25 @@ void VtkOutput::PrintOutput()
 /***********************************************************************************/
 /***********************************************************************************/
 
-void VtkOutput::WriteModelPartToFile(const ModelPart& rModelPart, const bool IsSubModelPart)
+void VtkOutput::WriteModelPartToFile(const ModelPart& rModelPart, const bool IsSubModelPart, const std::string& rOutputFilename)
 {
     Initialize(rModelPart);
 
+    std::string output_file_name = "";
+    if (rOutputFilename == "")
+    {
+        output_file_name = GetOutputFileName(rModelPart, IsSubModelPart);
+    }
+    else
+    {
+        if (mOutputSettings["save_output_files_in_folder"].GetBool())
```

Comment:
this is  duplicated now, I would prefer to do it in `GetOutputFileName`

~~~c++
    // Putting everything together
    std::string output_file_name;
    if (mOutputSettings["save_output_files_in_folder"].GetBool()) {
        output_file_name += mOutputSettings["folder_name"].GetString() + "/";
    }
    if (rOutputFilename != "") {
        output_file_name += rOutputFilename;
    } else {
        const std::string& r_custom_name_prefix = mOutputSettings["custom_name_prefix"].GetString();
        const std::string& r_custom_name_postfix = mOutputSettings["custom_name_postfix"].GetString();
        output_file_name += r_custom_name_prefix + model_part_name + r_custom_name_postfix + "_" + std::to_string(rank) + "_" + label + ".vtk";
    }
~~~

Reply:
or even clearer:
~~~c++
std::string VtkOutput::GetOutputFileName(const ModelPart& rModelPart, const bool IsSubModelPart)
{
    std::string output_file_name;
    if (mOutputSettings["save_output_files_in_folder"].GetBool()) {
        output_file_name += mOutputSettings["folder_name"].GetString() + "/";
    }
    
    if (rOutputFilename != "") {
        output_file_name += rOutputFilename;
    } else {
        const int rank = rModelPart.GetCommunicator().MyPID();
        std::string model_part_name;

        if (IsSubModelPart) {
            model_part_name = rModelPart.GetParentModelPart()->Name() + "_" + rModelPart.Name();
        } else {
            model_part_name = rModelPart.Name();
        }

        std::string label;
        std::stringstream ss;
        const std::string output_control = mOutputSettings["output_control_type"].GetString();
        if (output_control == "step") {
            ss << std::fixed << std::setprecision(mDefaultPrecision)<< std::setfill('0')
            << rModelPart.GetProcessInfo()[STEP];
            label = ss.str();
        } else if(output_control == "time") {
            ss << std::fixed << std::setprecision(mDefaultPrecision) << std::setfill('0')
            << rModelPart.GetProcessInfo()[TIME];
            label = ss.str();
        } else {
            KRATOS_ERROR << "Option for \"output_control_type\": " << output_control
                <<" not recognised!\nPossible output_control_type options "
                << "are: \"step\", \"time\"" << std::endl;
        }
        const std::string& r_custom_name_prefix = mOutputSettings["custom_name_prefix"].GetString();
        const std::string& r_custom_name_postfix = mOutputSettings["custom_name_postfix"].GetString();
        output_file_name += r_custom_name_prefix + model_part_name + r_custom_name_postfix + "_" + std::to_string(rank) + "_" + label + ".vtk";
    }

    return output_file_name;
}
~~~